source/Irrlicht/CGLTFMeshFileLoader.cpp

```diff
@@ -412,70 +428,53 @@ std::size_t CGLTFMeshFileLoader::MeshExtractor::getPositionAccessorIdx(
  * Type: VEC3 (Float)
  * ! Required: NO (Appears to not be, needs another pair of eyes to research.)
 */
-std::size_t CGLTFMeshFileLoader::MeshExtractor::getNormalAccessorIdx(
+std::optional<std::size_t> CGLTFMeshFileLoader::MeshExtractor::getNormalAccessorIdx(
 		const std::size_t meshIdx,
 		const std::size_t primitiveIdx) const
 {
-	const auto& attributes = m_model.meshes[meshIdx]
-		.primitives[primitiveIdx].attributes;
-	const auto result = attributes.find("NORMAL");
-
-	if (result == attributes.end()) {
-		return -1;
-	} else {
-		return result->second;
-	}
+	return m_model.meshes->at(meshIdx).primitives[primitiveIdx].attributes.normal;
 }
 
 /**
- * The index of the accessor that contains the NORMALs.
+ * The index of the accessor that contains the TEXCOORDs.
  * Documentation: https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#meshes-overview
  * Type: VEC3 (Float)
  * ! Required: YES (Appears so, needs another pair of eyes to research.)
 */
-std::size_t CGLTFMeshFileLoader::MeshExtractor::getTCoordAccessorIdx(
+std::optional<std::size_t> CGLTFMeshFileLoader::MeshExtractor::getTCoordAccessorIdx(
 		const std::size_t meshIdx,
 		const std::size_t primitiveIdx) const
 {
-	const auto& attributes = m_model.meshes[meshIdx]
-		.primitives[primitiveIdx].attributes;
-	const auto result = attributes.find("TEXCOORD_0");
-
-	if (result == attributes.end()) {
-		return -1;
-	} else {
-		return result->second;
-	}
+	const auto& texcoords = m_model.meshes->at(meshIdx).primitives[primitiveIdx].attributes.texcoord;
+	if (!texcoords.has_value())
+		return std::nullopt;
+	return texcoords->at(0);
 }
 
 /**
- * This is where the actual model's GLTF file is loaded and parsed by tinygltf.
+ * This is where the actual model's GLTF file is loaded and parsed by tiniergltf.
 */
-bool CGLTFMeshFileLoader::tryParseGLTF(io::IReadFile* file,
-		tinygltf::Model& model)
+std::optional<tiniergltf::GlTF> CGLTFMeshFileLoader::tryParseGLTF(io::IReadFile* file)
 {
-	tinygltf::TinyGLTF loader {};
-
-	// Stop embedded textures from making model fail to load
-	loader.SetImageLoader(nullptr, nullptr);
-
-	std::string err {};
-	std::string warn {};
-
-	auto buf = std::make_unique<char[]>(file->getSize());
-	file->read(buf.get(), file->getSize());
-
-        if (warn != "") {
-                os::Printer::log(warn.c_str(), ELL_WARNING);
-        }
-
-	if (err != "") {
-                os::Printer::log(err.c_str(), ELL_ERROR);
-		return false;
+	auto size = file->getSize();
+	auto buf = std::make_unique<char[]>(size + 1);
+	file->read(buf.get(), size);
+	// We probably don't need this, but add it just to be sure.
+	buf[size] = '\0';

```

Comment:
Yes, this is a buffer, not a null-terminated string. Don't add a null-terminator. What we should do is use a span to keep the buffer size together with the buffer.

Reply:
I based this on the https://github.com/open-source-parsers/jsoncpp/blob/master/example/readFromString/readFromString.cpp example, so I'm not sure whether jsoncpp requires it to be null-terminated. I would assume not, because it requires you to pass the end pointer.

The [docs](http://open-source-parsers.github.io/jsoncpp-docs/doxygen/class_json_1_1_char_reader.html#a65517004c4b5b1dc659ab966b3cea4cf) say "[t]he document must be a UTF-8 encoded string containing the document to read", so that doesn't really help :/

Reply:
The docs don't say it has to be null-terminated, so I don't think it has to be, but since I've had time to think about it more, I'm fine with leaving this in.